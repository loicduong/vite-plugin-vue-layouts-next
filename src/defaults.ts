import type { ResolvedOptions, UserOptions } from './types.js'
import process from 'node:process'

export function defaultImportMode(name: string) {
  if (process.env.VITE_SSG)
    return 'sync'

  return name === 'default' ? 'sync' : 'async'
}

export function resolveOptions(userOptions: UserOptions): ResolvedOptions {
  return Object.assign(
    {
      defaultLayout: 'default',
      layoutsDirs: 'src/layouts',
      pagesDirs: 'src/pages',
      extensions: ['vue'],
      exclude: [],
      wrapComponent: false,
      importMode: defaultImportMode,
    },
    userOptions,
  )
}
