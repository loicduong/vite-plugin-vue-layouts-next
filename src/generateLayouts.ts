import type { ResolvedConfig } from 'vite'
import type { FileContainer, ResolvedOptions } from './types.js'
import { resolve } from 'node:path'
import { getFilesFromPath } from './files.js'
import { getImportCode } from './importCode.js'
import getClientCode from './RouteLayout.js'
import { debug, normalizePath } from './utils.js'

export async function generateLayouts(layoutDirs: string[], options: ResolvedOptions, config: ResolvedConfig) {
  const container: FileContainer[] = []

  for (const dir of layoutDirs) {
    const layoutsDirPath = dir.substr(0, 1) === '/'
      ? normalizePath(dir)
      : normalizePath(resolve(config.root, dir))

    const _f = await getFilesFromPath(layoutsDirPath, options)
    container.push({ path: layoutsDirPath, files: _f })
  }

  const importCode = getImportCode(container, options)

  const clientCode = getClientCode(importCode, options)

  debug('Client code: %O', clientCode)
  return clientCode
}
