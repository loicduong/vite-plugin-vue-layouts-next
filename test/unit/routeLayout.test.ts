import type { ResolvedConfig } from 'vite'
import { describe, expect, it } from 'vitest'
import { resolveOptions } from '../../src/defaults.js'
import { generateLayouts } from '../../src/generateLayouts.js'
import { resolveDirs } from '../../src/utils.js'

/* describe('virtual module utilities', () => {
}) */

describe('createVirtualModuleCode', async () => {
  const root = './'
  const layoutsDirs = './layouts'
  const rConfig = { root } as ResolvedConfig

  it('generates code with default layout', async () => {
    const code = await generateLayouts(resolveDirs([
      './layouts',
    ], root), resolveOptions({
      layoutsDirs,
      defaultLayout: 'default',
      importMode: () => 'sync',
      // skipTopLevelRouteLayout: false,
      wrapComponent: false,
    }), rConfig)

    expect(code).toMatchSnapshot()
  })

  it('generates code with wrapped components', async () => {
    const code = await generateLayouts(resolveDirs([
      './layouts',
    ], root), resolveOptions({
      layoutsDirs,
      defaultLayout: 'default',
      importMode: () => 'async',
      // skipTopLevelRouteLayout: false,
      wrapComponent: true,
    }), rConfig)

    expect(code).toMatchSnapshot()
  })

  it('generates code with top level route layout skipping', async () => {
    const code = await generateLayouts(resolveDirs([
      './layouts',
    ], root), resolveOptions({
      layoutsDirs,
      defaultLayout: 'default',
      importMode: () => 'sync',
      // skipTopLevelRouteLayout: true,
      wrapComponent: false,
    }), rConfig)

    expect(code).toMatchSnapshot()
  })
})
