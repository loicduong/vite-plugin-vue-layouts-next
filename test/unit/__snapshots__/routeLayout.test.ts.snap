// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createVirtualModuleCode > generates code with default layout 1`] = `
"


export const layouts = {

}
export const createGetRoutes = (router, withLayout = false) => {
  const routes = router.getRoutes()
  if (withLayout) {
      return routes
  }
  return () => routes.filter(route => !route.meta.isLayout)
}

export function setupLayouts(routes) {
  function deepSetupLayout(routes, top = true) {
    return routes.map(route => {
      if (route.children?.length > 0) {
        route.children = deepSetupLayout(route.children, false)
      }
      
      const layout = route.meta?.layout ?? 'default'

      const skipLayout = top
        && !route.component
        && route.children?.find(r => (r.path === '' || r.path === '/') && r.meta?.isLayout);

      if (skipLayout) {
        return route
      }
          
      if (layout && layouts[layout]) {
        
/** @type {import('vue-router').RouteRecordRaw} */
const layoutRoute = {
        path: route.path,
        component: layouts[layout],
        meta: { ...route.meta, isLayout: true },
        // Handle root path specially to avoid infinite nesting
        children: top && route.path === '/' 
        ? [route] 
        : [{ ...route, path: '', meta: { ...route.meta, isLayout: false } }]
}
return layoutRoute

      }

      return route
    })
  }

    return deepSetupLayout(routes)

}
"
`;

exports[`createVirtualModuleCode > generates code with top level route layout skipping 1`] = `
"


export const layouts = {

}
export const createGetRoutes = (router, withLayout = false) => {
  const routes = router.getRoutes()
  if (withLayout) {
      return routes
  }
  return () => routes.filter(route => !route.meta.isLayout)
}

export function setupLayouts(routes) {
  function deepSetupLayout(routes, top = true) {
    return routes.map(route => {
      if (route.children?.length > 0) {
        route.children = deepSetupLayout(route.children, false)
      }
      
      const layout = route.meta?.layout ?? 'default'

      const skipLayout = top
        && !route.component
        && route.children?.find(r => (r.path === '' || r.path === '/') && r.meta?.isLayout);

      if (skipLayout) {
        return route
      }
          
      if (layout && layouts[layout]) {
        
/** @type {import('vue-router').RouteRecordRaw} */
const layoutRoute = {
        path: route.path,
        component: layouts[layout],
        meta: { ...route.meta, isLayout: true },
        // Handle root path specially to avoid infinite nesting
        children: top && route.path === '/' 
        ? [route] 
        : [{ ...route, path: '', meta: { ...route.meta, isLayout: false } }]
}
return layoutRoute

      }

      return route
    })
  }

    return deepSetupLayout(routes)

}
"
`;

exports[`createVirtualModuleCode > generates code with wrapped components 1`] = `
"

import { h } from 'vue'


export const layouts = {

}
const defineAsyncComponent = null
const defineComponent = null

export const createGetRoutes = (router, withLayout = false) => {
  const routes = router.getRoutes()
  if (withLayout) {
      return routes
  }
  return () => routes.filter(route => !route.meta.isLayout)
}

export function setupLayouts(routes) {
  function deepSetupLayout(routes, top = true) {
    return routes.map(route => {
      if (route.children?.length > 0) {
        route.children = deepSetupLayout(route.children, false)
      }
      
      const layout = route.meta?.layout ?? 'default'

      const skipLayout = top
        && !route.component
        && route.children?.find(r => (r.path === '' || r.path === '/') && r.meta?.isLayout);

      if (skipLayout) {
        return route
      }
          
      if (layout && layouts[layout]) {
        
if (!route.component) {
        return route
}
/** @type {import('vue-router').RouteRecordRaw} */
const wrappedRoute = {
        ...route,
        component: h('div', [
        h(layouts[layout].layout),
        h(layouts[layout].isSync ? defineComponent(() => route.component()) : defineAsyncComponent(() => route.component())),
        ]),
        meta: {
        ...route.meta,
        isLayout: true
        }
}
return wrappedRoute

      }

      return route
    })
  }

    return deepSetupLayout(routes)

}
"
`;
